### DISCUSSION TITLE
API Discussion: Cryptographic peer id (fedimint_api::PeerId) instead of sequential numbers (0, 1, 2, ...)

### DISCUSSION THREAD

**Discussion by yojoe** [fedimint/fedimint#5733] [General]
# TL;DR
I started researching (and hacking) on a peer-to-peer (P2P) based version of `fedimintd`, where nodes can be behind NAT routers and don't require a DNS name, static IP address configuration (or dyndns domain) for the `Federation Code` and to also cover the worst case scenario, where neither of the peers is able to leverage [UPnP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) or set up manual port forwarding. To enable the worst case, a technique called [holepunching](https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html) is required. Luckily there are pretty mature rust libraries like [libp2p](https://docs.rs/libp2p) who handle this stuff already, so there's little need to reinvent the wheel.
While trying to achieve this with minimal refactoring of the current upstream code (so I can eventually submit a pull request) and an optional feature (`cargo build -p fedimintd --features p2p`) I ended up wondering/questioning if the current approach of sequential integers for `fedimint_api::PeerId` will reach its limits sooner than later anyway and if it's time to start this discussion now? Reading the comments in https://github.com/fedimint/fedimint/pull/9 I get the impression the `PeerId` struct emerged along the way and was not an essential concept during the design phase of the fedimint API?

# Long explanation
Using cryptographic peer ids, the "peering" process between federation guardians (while creating the federation) and also the "peering" with / onboarding of federation members could be simplified a lot. The process could basically be reduced to an out-of-bounds exchange of the cryptographic peer ids and all the remaining configuration data could be fetched from the infrastructure (even connection endpoints). This way you should be able get rid of a lot of the (currently manual) configuration of listening addresses and ports, for example by using a distributed hash table DHT (like Kademlia) instead, where peers can discover (and update) their connection information dynamically. Whether or not you want to implement this now is a different discussion, but cryptographic peer ids instead of sequential integers (which are subjective to the local node) would be a necessary building block and has tremendous (refactoring) impact on the API, if not thoroughly thought of early on, as peer ids are needed almost everywhere. My main concern right now is, that sequential peer ids are currently used a lot (like `peer_id.to_usize()`) as indexes to access array values, setup port numbers for connection endpoints, indexes for the shares in `PublicKeySet` (`threshold-crypto`), etc.

# Analysis
To see how far I get with the current upstream code, I forked a local branch and introduced an optional [feature](https://doc.rust-lang.org/cargo/reference/features.html) to change the type of `fedimint_api::PeerId` not being a `u16` but a more complex structure - to be specific, I'm experimenting with `fedimint_api::PeerId` being a re-export of [libp2p::PeerId](https://docs.rs/libp2p/latest/libp2p/struct.PeerId.html), if and only if my rust feature is enabled during compile time.

Unfortunately, this requires a lot of changes to the existing code base as `fedimint_api::PeerId` is used almost everywhere and I end up with lots of conditional code like:

```rust
#[cfg(not(feature = "libp2p"))]
let tls_keys = peers
    .iter()
    .map(|peer| {
        let (cert, key) = gen_cert_and_key(&format!("peer-{}", peer.to_usize())).unwrap();
        (*peer, (cert, key))
    })
    .collect::<HashMap<_, _>>();

#[cfg(feature = "libp2p")]
let tls_keys = peers
    .iter()
    .map(|peer| {
        let (cert, key) = gen_cert_and_key(&format!("peer-{}", peer.to_base58())).unwrap();
        (*peer, (cert, key))
    })
    .collect::<HashMap<_, _>>();
```

It's certainly doable, but ugly and repetitive. But the majority of the changes to the existing codebase is related to the fact that the `PeerId::to_usize` is currently used a lot directly as an index to access various array values. To achieve what I want, this needs to be refactored to enumerating the peer set and using the natural order (iterator index) to access the respective array values. At least this refactoring works for both - sequential peer ids and cryptographic peer ids:

current upstream code example (numeric peer ids):
```rust
let mint_cfg = peers
            .iter()
            .map(|&peer| {
                ...
                    .map(|amount| (*amount, tbs_keys[amount].2[peer.to_usize()]))
                ...
```

refactored code example (numeric or cryptographic peer ids):
```rust
let mint_cfg = peers
            .iter()
            .enumerate()
            .map(|(idx, &peer)| {
                ...
                    .map(|amount| (*amount, tbs_keys[amount].2[idx]))
                ...
```
But still, while this would fix the existing code base with limited effort, sequential peer ids are tempting to be used as indexes again for new code written.

Other issues with sequential peer ids are that examples and tests use the peer id directly to derive hard coded port numbers etc. But this is mentioned in https://github.com/fedimint/fedimint/issues/304 already.

# Action item
Long story short, I suggest to re-consider the use of sequential peer ids for the Fedimint API and maybe jump to something like [libp2p::PeerId](https://docs.rs/libp2p/latest/libp2p/struct.PeerId.html), which is agnostic to the signature algorithm you want to use (Ed25519, Secp256k1, RSA) and uses self-describing [multihashes](https://multiformats.io/multihash/). Any thoughts?

**Comment by elsirion** [comment]
I'll think about it a bit more but this sounds like a great idea. Some comments:

* We'll still need an integer !=0 associated with each peer for threshold blind signature key/signature polynomial interpolation. But if we have access to all peers we can indeed just sort and enumerate them (that's what HBBFT does iirc).
* Re-using the transport encryption key hash seems to be a reasonable choice for the peer id (currently TLS, but could be something else).
* I'm not sure yet if pulling in libp2p is a good idea, but since the transport is modular this could be a separate crate in the workspace and only used optionally. It's definitely useful, we just need to reduce dependencies eventually. This would mean we need our own cryptographic peer id type.

You seem to have put a lot of thought into this already, would you like to take point on this?

**Comment by yojoe** [comment]
> * Re-using the transport encryption key hash seems to be a reasonable choice for the peer id (currently TLS, but could be something else).

Usually it's the other way around, (ephemeral) transport encryption keys are derived/authenticated from a long-lived peer id. Especially if there are multiple transports a node supports and peers can select/negotiate. Here's a libp2p example code I'm playing with for authentication of a TCP transport secured with Noise (instead of TLS) which is generated on node startup and authenticated using the `local_keypair`(loaded from config) which is used as the PeerId throughout the application:

```rust
    // setup encrypted transports using mplex streams
    // plain TCP (TLS(rustls), Noise)
    // `tokio`-based TCP transport with libp2p-dns support
    let transport = dns::TokioDnsConfig::system(TokioTcpTransport::new(GenTcpConfig::default().nodelay(true))).unwrap()
        .upgrade(upgrade::Version::V1)

        // setup Noise keypair
        .authenticate(
            // TODO: NoiseAuthenticated::xx` constructor not yet released, will be available with libp2p-noise 0.40.0
            // noise::NoiseAuthenticated::xx(&local_keypair)
            {
                let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
                    .into_authentic(&local_keypair)
                    .expect("Signing libp2p-noise static DH keypari failed.");

                noise::NoiseConfig::xx(noise_keys).into_authenticated()
            }
        )

        // use mplex streams
        .multiplex(mplex::MplexConfig::new())

        .boxed();
```

**Comment by yojoe** [comment]
> * I'm not sure yet if pulling in libp2p is a good idea, but since the transport is modular this could be a separate crate in the workspace and only used optionally.

That's what I thought would be best.

> It's definitely useful, we just need to reduce dependencies eventually. This would mean we need our own cryptographic peer id type.

Makes sense. But compatibility with [libp2p::core::PeerId](https://docs.rs/libp2p/latest/libp2p/core/struct.PeerId.html) still makes sense so that you could easily convert to/from `fedimint-api::PeerId` to `libp2p::core::PeerId` including [multihash](https://docs.rs/multihash/latest/multihash/) usage for peer ids. This helps a lot to make applications "future-proof":

> Multihash is a protocol for differentiating outputs from various well-established hash functions, addressing size + encoding considerations. It is useful to write applications that future-proof their use of hashes, and allow multiple hash functions to coexist.

**Comment by jkitman** [comment]
@yojoe I'm not sure if replacing `fedimint_api::PeerId` is the right approach, although the name might be confusing and long-term perhaps we can refactor it.  As you might have found out we depend on it being a 0-indexed list of integers in many places making for a messy feature flag.

Our current trust model for guardians requires that they share network addresses, a base port, and TLS certs securely out-of-band.  Once authenticated connections are established between all peers they can run key generation algorithms for the threshold keys they need.  We just merged the code for this a few days ago.

I'd suggest we could replace the tls key/certs found in `ServerConfig` with your cryptographic "long-lived peer id" from which we could derive keys for network authentication (whether TLS or P2P).

Feel free to reach out over discord if you want to chat about it.

**Comment by EthnTuttle** [comment]
If this is ever circled back to, there will be relevant UI updates now. Just commenting here for awareness and so I get notifications.

**Comment by jkitman** [comment]
Given our assumption that the federation id is the single source of truth for all clients (e.g. publicly posted on well-known authenticated forums), the public key share of each guardian can authenticate messages from each guardian independently.

However, this is unrelated to a non-DNS method of discovery, which is more the concept of this issue.  Perhaps we should close this issue and open a new one.

**Comment by EthnTuttle** [comment]
So would the public key share be a unique identifier for a specific guardian? And if so, would that ever be exposed via the API?

I came across this issue when looking for a way to concretely identify what Guardian a user is during setup. I settled on using `our_current_id` https://github.com/fedimint/fedimint/blob/153e4af5f1af84c16ea1240669512f121f0d85ec/fedimint-core/src/admin_client.rs#L232 but it seemed this discussion proposed a move away from sequential numbers.

**Comment by jkitman** [comment]
I think `PeedId` is okay as an identifier especially in the UI.
